salesagility/SuiteCRM
# Fix basis for storing and processing information with regards to GDPR.

## Description



This will add in the ability to track the basis for which the business has permission to store a users information and process this. 

## Motivation and Context

To become fully compliant with GDPR.

## How To Test This


Pre-Testing Setup
1) Ensure that the fields are added into the default by going into: Admin -> Studio -> (The desired 
    Module) -> Edit  view / Detail view -> Add the fields into the display,
2) the fields are: Lawful basis, Lawful basis source and Date Reviewed,
3) Have the email settings set to Confirm Opt In
    3.1) Navigate to the admin panel,
    3.2) Select 'Email Settings',
    3.3) Set the 'Opt In' settings to 'Confirmed Opt In'.

Testing addition fields - 1.1
1) Enter the contacts module,
2) Create a contact ,
3) See on the edit view that there are 3 new fields, these are: Lawful Basis, Lawful Basis Source and 
    Date reviewed,
4) Once the contact has been created, see that these new fields are also visible in detail view,
5) Carry out the same steps as above with regards to the Leads module and targets module.

Testing for audit-able fields - 2.1
1) Navigate to the modules as mentioned above, Contacts, Leads and Targets,
2) Select a contact/lead/target, if ones exists. If not, create one,
3) Enter the detail view of the record,
4) Click the actions tab, and select 'View Change Log',
5) See that the 3 new fields are available in the change log.
 
Testing for ability to mass update fields - 3.1 
1) Navigate to the modules as mentioned above, Contacts, Leads and Targets,
2) Ensure that you are in the list view,
3) Select the checkbox for one or more of the records on the left hand side or select all at the top/bottom 
    of the list,
4) Select the bulk action drop down and select mass update,
5) Scroll to below the list view and see that the 3 new fields are visible and able to be updated.

Testing the persistent through the life-cycle of the record - 4.1
1) Create a Lead within the system,
2) Add some data into the new fields, Lawful basis source, Date reviewed and lawful basis,
3) Once the lead has been created, convert the lead,
    -> Enter the detail view of the lead 
    -> select the actions button and click convert lead
    -> Fill out the form which is loaded, once happy, click save at the top or bottom,
4) Once this has successfully completed, navigate to the contacts module,
5) Find the newly created record,
6) Check the record to see if the new fields have been carried across.

Testing the import of the new fields - 5.1
1) Navigate to the modules as mentioned above, Contacts, Leads and Targets,
2) Using the side bar, select the import button,
3) From here, download the import template,
4) Search through the template to ensure that the new fields are there,
5) Try importing a record by filling out the form,
6) Once it is completed, save the form and continue with the import,
7) Upon completion of the import, navigate to the new record and look to see if the new fields have 
    been populated.

Testing the WebToLead form submission - 6.1
1) Create a campaign withing the system,
    1.1) Navigate to campaigns module,
    1.2) Select Create Campaign,
    1.3) Create a campaign, e.g. Email,
2) Once there is a campaign, select Create Person Form,
3) For testing, drag into the column; Last Name and Email Address,
4) Select next,
5) Ensure that the Related Campaign has been selected,
6) Select Generate Form,
7) Click on the Web To Person Form link,
8) Have the Confirm opt in settings (Inside the email settings page) disabled,
9) Fill in the form and tick the Opt In checkbox,
10) Once the form has been submitted, see the Leads list view and see that the Lawful Basis and Lawful 
      Basis source has not been populated,
11) Repeat step 7
12) Have the confirm opt in settings this time set to Confirmed opt in,
13) Fill in the form again, don't select the opt in check box,
14) Once the form has been submitted, see the Leads list view and see that the Lawful Basis and 
      Lawful Basis source has not been populated,
15) Repeat step 7
16) Have the confirm opt in settings this time set to Confirmed opt in,
17) Fill in the form again, select the opt in check box,
18) Once the form has been submitted, see the Leads list view and see that the Lawful Basis and 
      Lawful Basis source has been populated with the following;
      Lawful Basis source: Website,
      Lawful Basis: Consent.      

Testing the Confirm Op In form submission - 7.1
1) Have a Contact/Lead/Target within the system which has the email status of 'Opt in' ,
    1.1) This can be done by navigating to the edit view,
    1.2) Enter an email address and have the 'Opt In' checkbox selected,
2) Send a Confirm Opt In email to the record,
    2.1) This can be done by Navigating to the List view,
    2.2) Selecting the checkbox for the record and clicking the 'Bulk Actions' button,
    2.3) Select the 'Send Confirm Opt In' Option,
3) From within the email, click the 'opt in' link,
4) Navigate back to the list view within the CRM to view the record,
5) The newly opted in record should now have the following values:
    5.1) Lawful Basis: Consent,
    5.2) Lawful Basis Source: Email,
    5.3) Date Reviewed: The date that the link was clicked in the users format.

## Types of changes


### Final checklist
